### PR TITLE
Hw9

### PR DESCRIPTION
**Pull request by AscentionOne**
Hi, Jenn week 9 homework is ready for review. Thank you!

### CODE REVIEW

File: app/src/test/java/com/kenchen/capstonenewsapp/NewsRepositoryImpTest.kt [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        every { mockNetworkStatusChecker.isOnWifiConnection() } returns true
+        coEvery { mockRemoteApi.getTopHeadlinesByCountry("us") } returns listOf(dummyArticle)
+
+        coEvery { mockArticleDao.clearArticles() } just runs
+        coEvery { mockArticleDao.addArticles(any()) } just runs
+
+        runBlocking {
+            val flowArticleState = newsRepository.getArticles()
+
+//            println(flowArticleState.take(3).toList())
```
Comment (zlippard) [inline app/src/test/java/com/kenchen/capstonenewsapp/NewsRepositoryImpTest.kt:162]: I never did get a chance to look in to this issue with `take()` and `drop()` 😅 Overall though, you have really good tests here for your repo that do a pretty thorough job of testing different edge cases!

File: app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+
+        // FIXME: can we put every in set up before test start?
+//        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+//        viewModel = NewsViewModel(mockRepository)
```
Comment (zlippard) [inline app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt:52]: This _should_ work

File: app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        viewModel.toggleDataUsage()
+
+        coVerify(exactly = 1) { mockRepository.toggleDataUsage() }
+    }
+}
+
+/**
+ * Question:
+ * 1. how do I test the function that returns Unit (nothing)
```
Comment (zlippard) [inline app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt:141]: Good questions!

For this one, you would want to ensure and verify that underlying methods are being called.

For example if I have the following:

```kotlin
class A {
    fun aFunction() { ... }
}

class B(val a: A) {
    fun bFunction() {
        a.aFunction()
    }
}
```

Then your test would do this:

```kotlin
val mockA = mockk<A>()
every { mockA.aFunction() } just runs

val sut = B(a)

sut.bFunction()

verify(exactly = 1) { mockA.aFunction() }
```

File: app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+ * 2. what is the difference between black-box and white-box testing
```
Comment (zlippard) [inline app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt:142]: Black box testing involves testing that has _no understanding_ of the inner workings and software components of the system being tested.  Appium is a good example of a black box test framework for mobile platforms.  You write your test code to "simulate" a user interaction and it runs those commands on the app. For example, tap a button, enter email, tap login button, enter password, etc.  These tests are typically written by testers or QA.

White box testing means you have a good understanding of the software and are testing individual pieces of the inner software.  Integration testing and unit testing are good examples here.  These tests are written by the software developers themselves.

File: app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+ * 3. can we put all the code in runBlocking ?
```
Comment (zlippard) [inline app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt:143]: For the most part, yes.  There's a `runTest` block that you can call as well, which arguably is a bit better since it runs all coroutine on a "test" dispatcher and not "main".

**Review by zlippard — APPROVED** [review]
